python/pyarrow/array.pyx

```diff
@@ -45,16 +46,40 @@ cdef _pandas():
     return pd
 
 
+cdef maybe_coerce_datetime64(values, dtype, DataType type,
```

Comment:
Either `type_` or `dtype` would be ok for me but as it's internally used, we can keep it. I wasn't 100% sure about the implications of using a reserved keyword.